src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs

```diff
@@ -12,18 +12,57 @@ namespace Microsoft.PowerShell.Commands
     [OutputType(typeof(System.IO.FileInfo))]
     public class NewTemporaryFileCommand : Cmdlet
     {
+        private string _extension = ".tmp";
+
+        /// <summary>
+        /// Specify a different file extension other than the default one, which is '.tmp'. The period in this parameter is optional
+        /// </summary>
+        [Parameter(Position = 0)]
+        [ValidateNotNullOrEmpty]
+        public string Extension
+        {
+        get { return _extension; }
+        set { _extension = value; }
+        }
+
         /// <summary>
         /// returns a TemporaryFile
         /// </summary>
         protected override void EndProcessing()
         {
             string filePath = null;
-            string tempPath = System.Environment.GetEnvironmentVariable("TEMP");
-            if (ShouldProcess(tempPath))
+            string tempPath = Path.GetTempPath();
+            int attempts = 0;
+            bool temporaryFilePathFound = false;
+            while(attempts++ < 10 && !temporaryFilePathFound)
+            {
+                string fileName = Path.GetRandomFileName();
+                fileName = Path.ChangeExtension(fileName, _extension);
+                filePath = Path.Combine(tempPath, fileName);
+                if(!File.Exists(filePath))
```

Comment:
I was aware of this when writing it but did not consider handling it specifically since the global catch would have still handled it and the probability of having a clash of a random file name in such a short time period is really low. However, I liked your idea and as your approach does not seem to have a downside, I implemented your suggestion.